runners/jstorm/src/main/java/org/apache/beam/runners/jstorm/translation/JStormPipelineTranslator.java

```
@@ -0,0 +1,121 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(JStormPipelineTranslator.class);
+
+  private final TranslationContext context;
+  private int depth = 0;
+
+  public JStormPipelineTranslator(TranslationContext context) {
+    this.context = context;
+  }
+
+  public void translate(Pipeline pipeline) {
+    pipeline.traverseTopologically(this);
+  }
+
+  @Override
+  public CompositeBehavior enterCompositeTransform(TransformHierarchy.Node node) {
+    LOG.debug(genLogPrefix(this.depth) + "enterCompositeTransform- " + node);
+    this.depth++;
+
+    // check if current composite transforms need to be translated.
+    // If not, all sub transforms will be translated in visitPrimitiveTransform.
+    PTransform<?, ?> transform = node.getTransform();
+    if (transform != null) {
+      TransformTranslator translator = TranslatorRegistry.getTranslator(transform);
+
+      if (translator != null && applyCanTranslate(transform, node, translator)) {
+        applyTransform(transform, node, translator);
+        LOG.debug(genLogPrefix(this.depth) + "translated-" + node);
+        return CompositeBehavior.DO_NOT_ENTER_TRANSFORM;
+      }
+    }
+    return CompositeBehavior.ENTER_TRANSFORM;
+  }
+
+  @Override
+  public void leaveCompositeTransform(TransformHierarchy.Node node) {
+    this.depth--;
+    LOG.debug(genLogPrefix(this.depth) + "leaveCompositeTransform- " + node);
+  }
+
+  @Override
+  public void visitPrimitiveTransform(TransformHierarchy.Node node) {
+    LOG.debug(genLogPrefix(this.depth) + "visitPrimitiveTransform- " + node);
+
+    if (!node.isRootNode()) {
+      PTransform<?, ?> transform = node.getTransform();
+      TransformTranslator translator = TranslatorRegistry.getTranslator(transform);
+      checkState(
+          translator != null && applyCanTranslate(transform, node, translator),
+          "The primitive transform " + transform + " is currently not supported.");
```

I think if it is unsupported you probably want to throw `UnsupportedOperationException`. But the translator will probably know why it cannot translate, so you might want to use an interface where it can return a reason.